### ISSUE TITLE
incorrect FFT results for Fortran-order arrays?

### ISSUE THREAD

**Issue by leofang** [cupy/cupy#1934]
* Conditions (from `python -c 'import cupy; cupy.show_config()'`)

Tested in two environments with different CuPy versions:
```bash
CuPy Version          : 4.4.1
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 9010
CUDA Driver Version   : 9010
CUDA Runtime Version  : 9010
cuDNN Build Version   : 7102
cuDNN Version         : 7102
NCCL Build Version    : 2115
```
and (this CuPy is built from the latest master branch)
```
CuPy Version          : 6.0.0b1
CUDA Root             : /usr/local/cuda
CUDA Build Version    : 9010
CUDA Driver Version   : 9010
CUDA Runtime Version  : 9010
cuDNN Build Version   : None
cuDNN Version         : None
NCCL Build Version    : None
```
* Code to reproduce
```python
import numpy as np
import cupy as cp

AXES=[(0,), (1,), (2,), (0,1), (1,2), (0,2), (0,1,2)]
a_np = np.random.random((3,4,5))+1j*np.random.random((3,4,5))

print("In C order:")
a_np = np.ascontiguousarray(a_np)
a_cp = cp.asarray(a_np)
a_cp = cp.ascontiguousarray(a_cp)
assert np.allclose(cp.asnumpy(a_cp), a_np)
for axes in AXES:
    result_np = np.fft.fftn(a_np, axes=axes)
    result_cp = cp.fft.fftn(a_cp, axes=axes)
    print(axes, ":", np.allclose(cp.asnumpy(result_cp), result_np))

print("\nIn F order:")
a_np = np.asfortranarray(a_np)
a_cp = cp.asarray(a_np)
a_cp = cp.asfortranarray(a_cp)
assert np.allclose(cp.asnumpy(a_cp), a_np)
for axes in AXES:
    result_np = np.fft.fftn(a_np, axes=axes)
    result_cp = cp.fft.fftn(a_cp, axes=axes)
    print(axes, ":", np.allclose(cp.asnumpy(result_cp), result_np))
```
* Error messages, stack traces, or logs

The outputs from both environments are identical:
```bash
In C order:
(0,) : True
(1,) : True
(2,) : True
(0, 1) : True
(1, 2) : True
(0, 2) : True
(0, 1, 2) : True

In F order:
(0,) : False
(1,) : True
(2,) : False
(0, 1) : True
(1, 2) : False
(0, 2) : False
(0, 1, 2) : True
```
But it's expected to be `True` for all of the axes choices. It seems to me the bug is not introduced by the recent changes in adding support for cuFFT plans (#1669, #1745, #1746) but by something much older. For now I have not yet tracked down the problem, will update here if I find it. I hope I didn't do something stupid in the test...

Thanks.

**Comment by grlee77** [comment]
If I try from current master so that it uses the recently merged `PlanNd` for most of these cases, the result is better.  Only one of the 1D cases still fails: 

```
In C order:
(0,) : True
(1,) : True
(2,) : True
(0, 1) : True
(1, 2) : True
(0, 2) : True
(0, 1, 2) : True

In F order:
(0,) : True
(1,) : True
(2,) : False
(0, 1) : True
(1, 2) : True
(0, 2) : True
(0, 1, 2) : True
```
So, only one of the cases not using the n-dimensional plans fails.

**Comment by leofang** [comment]
@grlee77 but in my second environment I also used the most recent master...

**Comment by grlee77** [comment]
hmm. I tried on a system with CUDA 8.0, but can try later on a different system with CUDA 9.2 to see if I get the same result.

**Comment by fiarabbit** [comment]
I could reproduce the situation with @grlee77 .
```
CuPy Version          : 5.1.0
CUDA Root             : /home/user/.local/cuda/cuda-9.2
CUDA Build Version    : 9020
CUDA Driver Version   : 10000
CUDA Runtime Version  : 9020
cuDNN Build Version   : 7401
cuDNN Version         : 7401
NCCL Build Version    : 2307
```

```
In C order:
(0,) : True
(1,) : True
(2,) : True
(0, 1) : True
(1, 2) : True
(0, 2) : True
(0, 1, 2) : True

In F order:
(0,) : True
(1,) : True
(2,) : False
(0, 1) : True
(1, 2) : True
(0, 2) : True
(0, 1, 2) : True
```

**Comment by fiarabbit** [comment]
```
>>> np.fft.fftn(a_np, axes=(2,))
array([[[ 2.46078385e+00+1.88796109e+00j,
          9.51205057e-02+1.43065266e+00j,
         -3.55468543e-01-1.12233921e-01j,
         -3.53076587e-02-3.98434086e-01j,
         -1.25717559e-01-2.85957294e-01j],
        [ 3.12475909e+00+2.73135510e+00j,
          2.23879250e-02+2.40119143e-01j,
         -9.92032930e-01-6.29827149e-01j,
         -2.84723469e-01+3.71321698e-01j,
          1.68853124e-01+4.19031123e-01j],
        [ 2.08852096e+00+1.49318454e+00j,
         -7.38701892e-02+4.42665685e-01j,
          2.91638815e-01-1.14513630e-02j,
         -2.61975531e-01+3.34646825e-01j,
         -1.64935141e-01-5.63465473e-01j],
        [ 2.88075254e+00+1.95054050e+00j,
          3.74160443e-01+5.49911074e-01j,
         -7.73115446e-01+3.48598437e-01j,
          1.18277491e+00+6.71804041e-01j,
         -1.26706176e-01-6.11082932e-01j]],

       [[ 2.15179300e+00+1.62143963e+00j,
          2.39719708e-01-8.10114762e-02j,
          9.88790365e-02-3.12789254e-03j,
          2.42204828e-01-4.86283458e-01j,
          1.56108631e+00-5.88554238e-01j],
        [ 1.98132352e+00+1.79341178e+00j,
          5.46903263e-01+1.28323730e+00j,
          1.83976485e-03+8.28757525e-01j,
          4.90844059e-01+4.34296855e-01j,
          2.93667693e-01+4.33292967e-01j],
        [ 1.99385498e+00+2.90153023e+00j,
         -3.85685506e-01+1.88187842e-01j,
          6.66126471e-01-5.32218615e-01j,
         -7.92413404e-01+1.92601702e+00j,
         -5.65867342e-01+3.64509129e-01j],
        [ 3.02235203e+00+2.59009213e+00j,
          1.48846977e+00-1.93147757e-01j,
         -6.10088979e-01-4.97970298e-01j,
          2.43785690e-01+2.56956771e-01j,
         -4.37519507e-01+2.72983510e-01j]],

       [[ 2.80359502e+00+3.38414518e+00j,
         -3.21096224e-01+1.06575883e+00j,
          8.33630268e-01-6.78303948e-02j,
         -2.01329937e-01-4.15976642e-01j,
          9.46232109e-01+1.38270553e-01j],
        [ 2.27438023e+00+3.20508304e+00j,
         -1.36691432e+00+6.03338014e-01j,
         -1.05793764e-01-3.56847465e-01j,
          3.70337201e-01+1.20959324e+00j,
          6.03437330e-01+1.65718363e-01j],
        [ 3.48600978e+00+2.26097152e+00j,
         -7.41542122e-01-1.31081898e-01j,
          7.57379383e-01+9.72317614e-01j,
         -4.23510439e-01+1.70645610e-01j,
          7.09978676e-01+5.19332115e-01j],
        [ 2.18919103e+00+1.20140375e+00j,
          9.36938673e-02-9.00900422e-01j,
         -4.24450075e-01+1.05329339e+00j,
          1.42724610e-02-4.41502047e-01j,
         -5.13518772e-01-1.74408410e-01j]]])
>>> cp.fft.fftn(a_cp, axes=(2,))
array([[[ 3.14958935+2.99876298j, -0.8145503 -0.76681413j,
         -1.13783163+0.37398092j,  0.24554629-0.16495145j,
          0.59665689+0.08101013j],
        [ 2.37940447+3.61448942j, -0.17811448+0.50511013j,
         -0.57426526+0.40283286j,  0.11287379+1.10558698j,
          0.03554817-0.8011342j ],
        [ 1.8193964 +1.9727325j ,  0.21036538+0.33140965j,
         -0.29260246-0.14200319j,  1.20836395+0.11466159j,
          0.76147574+0.15211381j],
        [ 2.55169768+1.80468427j,  0.2334717 +0.79088348j,
          0.52282729+0.48464239j,  0.54353315-0.54485047j,
          1.12385957+0.05377661j]],

       [[ 2.68544496+1.45489215j,  0.03934904-0.81423812j,
         -0.20358611+0.61506401j,  1.29666753-0.2504779j ,
         -0.46435544-0.23533806j],
        [ 1.50501824+1.63154119j, -0.23960887-0.47327695j,
         -0.6217385 +0.04839527j,  0.24971992+0.14388209j,
          0.06810234+0.02098277j],
        [ 2.01048779+2.40210571j, -1.67831526-0.29328056j,
          0.80039146+0.2609849j , -0.12060811-1.12265878j,
          0.35838112-0.02935665j],
        [ 2.82424882+2.3155991j ,  0.33366753-0.04611356j,
          0.11918179+0.35409729j, -0.11647967+1.06542673j,
          1.09012241-0.6544482j ]],

       [[ 3.38782092+2.86785799j, -0.93523025-0.72691401j,
         -0.04721436-1.43616074j, -0.63866913-0.24217274j,
         -0.18874183-0.28151447j],
        [ 2.92412179+1.58766059j, -0.35475136+0.19613029j,
          0.060577  -0.98702362j,  0.24609693+0.37665575j,
          0.92090716-1.01263601j],
        [ 2.7953785 +3.09628187j,  0.55427172-0.10574524j,
          0.990239  +0.82370154j, -1.04910615+0.39704805j,
          0.28117685+0.53329914j],
        [ 2.4247071 +1.27451071j, -0.20008608-0.39174619j,
         -0.29669212+0.12177748j, -0.67370004-0.06350937j,
         -1.06485777-0.02498758j]]])
```

**Comment by fiarabbit** [comment]
I found the reason & brief solution. In short, in (0,), (1,), (2,) conditions, `cupy.fft.fft._fft` is called, which suppose c_contiguous array, while in other conditions `cupy.fft.fft._fftn` is called, which check whether it is c_contigous or f_contiguous
```
import numpy as np
import cupy as cp

AXES=[(0,), (1,), (2,), (0,1), (1,2), (0,2), (0,1,2)]
a_np = np.random.random((3,4,5))+1j*np.random.random((3,4,5))
a_np = np.asfortranarray(a_np)
a_cp = cp.asarray(a_np)
a_cp = cp.asfortranarray(a_cp)
assert np.allclose(cp.asnumpy(a_cp), a_np)

from cupy.cuda import cufft
from cupy.fft.fft import _fft, _fftn

wrong_cp = cp.fft.fftn(a_cp, axes=(2,))
correct_cp = cp.fft.fftn(a_cp.copy(), axes=(2,))

print(np.allclose(cp.asnumpy(wrong_cp), cp.asnumpy(_fft(a_cp, None, (2,), None, cufft.CUFFT_FORWARD))))  # True
print(np.allclose(cp.asnumpy(correct_cp), cp.asnumpy(_fftn(a_cp, None, (2,), None, cufft.CUFFT_FORWARD))))  # True

#Solution: just transpose it
print(np.allclose(cp.asnumpy(correct_cp.transpose(2, 1, 0)), cp.asnumpy(_fft(a_cp.transpose(2, 1, 0), None, (0,), None, cufft.CUFFT_FORWARD))))  # True
```
Because `cupy.fft.fft._fftn` is very recently introduced (which has a good wrapper), I wish someone will write a good wrapper for `cupy.fft.fft._fft` too.

**Comment by leofang** [comment]
@fiarabbit Note that your CuPy is v5.1.0, and the recent changes made to `cupy.fft.fft._fftn` are not incorporated in your installation yet, so your test just confirms my suspicion further that the bug is buried deeply in the early implementations (perhaps earlier than v4.4.1), at least irrelevant to @grlee77's `cupy.fft.fft._fftn` that supports cuFFT plans.

**Comment by leofang** [comment]
Also,

> #Solution: just transpose it

For me this is not a reliable solution from both viewpoints of library development and user experience. cuFFT should be able to handle all these situations correctly and trivially. I feel that the problem could be we compute the strides for Fortran arrays wrong, but I have not verified my guess yet (working on something urgent).

**Comment by leofang** [comment]
@grlee77 @fiarabbit I think I spotted the inconsistency among our outputs. The reason is that I was using NumPy v1.14.2 from conda. After updating to v.1.15.4 using `conda update numpy`, I got the same output as yours. So part of the bug is in my old NumPy, and the rest is addressed in @grlee77's PR #1944. 

The extra check I did is to do a consistency check: the order of the input array (`'C'` or `'F'`) shouldn't matter when we apply `np.fft.fftn` to it, but I saw different outputs. On the other hand, `cp.fft.fftn` gave identical results except for `axes=(2,)`, which is the problematic case you saw.

**Comment by grlee77** [comment]
> I think I spotted the inconsistency among our outputs. The reason is that I was using NumPy v1.14.2 from conda

Okay, that makes sense.  conda's default channel replaces numpy's fftpack with `mklfft` which had a bug with Fortran-ordered input that was only recently fixed:
https://github.com/IntelPython/mkl_fft/pull/30
numpy recently expanded their tests to check for this as well:
https://github.com/numpy/numpy/pull/12702

**Comment by leofang** [comment]
Just a bit digression from CuPy: how come this bug could live for so long in `mkl_fft`...

**Comment by leofang** [comment]
I see. Thanks for the update @fiarabbit! I think you were right on the contiguity issue, which was to be fixed in #1944.
